poc/tenant-controller/pkg/controllers/tenants/controller.go

```diff
@@ -67,63 +74,174 @@ func (c *Controller) Run(ctx context.Context) error {
 }
 
 func (c *Controller) createTenant(obj *tenantsapi.Tenant) {
-	glog.Info("createTenant: %#v", obj)
-	syncRBACForTenantCRD(obj)
+	glog.V(2).Infof("createTenant: %#v", obj)
+	c.syncRBACForTenant(obj)
+	for _, nsReq := range obj.Spec.Namespaces {
+		c.createNamespaceForTenant(obj, &nsReq)
+	}
+	glog.V(2).Infof("createTenant completed: %#v", obj)
+}
 
-	// TODO Add later ... sanity checks to ensure namespaces being requested are valid and not already assigned to another tenant
+func (c *Controller) updateTenant(old, obj *tenantsapi.Tenant) {
+	glog.V(2).Infof("updateTenant: %#v", obj)
+	c.syncRBACForTenant(obj)
+	// sort namespaces in old and new tenants to find out which ones
+	// to be created and which ones to be deleted.
+	oldNsList := make([]string, len(old.Spec.Namespaces))
+	for i, ns := range old.Spec.Namespaces {
+		oldNsList[i] = ns.Name
+	}
+	sort.Strings(oldNsList)
+	nsList := make([]*tenantsapi.TenantNamespace, len(obj.Spec.Namespaces))
+	for i := range obj.Spec.Namespaces {
+		nsList[i] = &obj.Spec.Namespaces[i]
+	}
+	sort.Slice(nsList, func(i, j int) bool {
+		return strings.Compare(nsList[i].Name, nsList[j].Name) < 0
+	})
+	var i, j int
+	for i < len(oldNsList) && j < len(nsList) {
+		if res := strings.Compare(oldNsList[i], nsList[j].Name); res == 0 {
+			c.syncNamespaceForTenant(obj, nsList[j])
+			i++
+			j++
+		} else if res < 0 {
+			c.deleteNamespaceForTenant(obj, oldNsList[i])
+			i++
+		} else {
+			c.createNamespaceForTenant(obj, nsList[j])
+			j++
+		}
+	}
 
-	namespace := corev1.Namespace{}
+	for ; j < len(nsList); j++ {
+		c.createNamespaceForTenant(obj, nsList[j])
+	}
+	for ; i < len(oldNsList); i++ {
+		c.deleteNamespaceForTenant(obj, oldNsList[i])
+	}
+}
+
+func (c *Controller) deleteTenant(obj *tenantsapi.Tenant) {
+	glog.V(2).Infof("deleteTenant: %#v", obj)
 
-	for n := range obj.Spec.Namespaces {
-		namespace.ObjectMeta.Name = obj.Spec.Namespaces[n].Name
-		//Create namespace
-		c.k8sclient.CoreV1().Namespaces().Create(&namespace)
+	// TODO with OwnerReferences, no extra work is needed in deletion,
+	// remove the following code later.
 
-		glog.Info("Created namespace: %s", obj.Spec.Namespaces[n].Name)
+	c.deleteRBACForTenant(obj)
+	for _, nsReq := range obj.Spec.Namespaces {
+		c.deleteNamespaceForTenant(obj, nsReq.Name)
 	}
-
-	glog.Info("createTenant completed: %#v", obj)
-	// TODO create RBAC inside namespace
+	glog.Infof("deleteTenant completed: %#v", obj)
 }
 
-func (c *Controller) updateTenant(old, obj *tenantsapi.Tenant) {
-	glog.Info("updateTenant: %#v", obj)
-	syncRBACForTenantCRD(obj)
-	// TODO sync namespace
-	// TODO sync RBAC inside namespace
+func (c *Controller) createNamespaceForTenant(tenant *tenantsapi.Tenant, nsReq *tenantsapi.TenantNamespace) error {
+	if err := c.ensureNamespaceExists(tenant, nsReq.Name); err != nil {
```

Comment:
Minor comment: The names of the functions ensureNamespaceExists and syncNamespaceForTenant are slightly misleading.  The first one is not ensuring but actually creating the namespace (maybe a name like createNamespaceForTenant is something to consider ?).  The second one is attaching rbac for this namespace (so maybe a name like syncRbacForTenant or syncPropertiesForTenant could be considered ?).  Anyway its a minor comment.